proposals/nnnn-move-function.md

```diff
@@ -0,0 +1,455 @@
+the actual lifetime of `y` to release it after its last use, there isn't
+a strong guarantee that it will. Even if the optimizer does what we want,
+programmers modifying this code in the future
+may introduce new references to `y` that inadvertently extend its lifetime
+and break our attempt to keep `x` unique.
+
+Swift-evolution pitch thread: [https://forums.swift.org/t/pitch-move-function-use-after-move-diagnostic](https://forums.swift.org/t/pitch-move-function-use-after-move-diagnostic)
+
+
+## Proposed solution: Move Function + "Use After Move" Diagnostic
+
+That is where the `move` function comes into play. The `move` function is a new
+generic stdlib function that when given a local let, local var, or parameter
+argument provides a compiler guarantee to the programmer that the binding will
+be unable to be used again locally. If such a use occurs, the compiler will emit
+an error diagnostic. We can modify the previous example to use `move` to
+explicitly end the lifetime of `y` when we're done with it:
+
+```
+func test() {
+  var x: [Int] = getArray()
+  
+  // x is appended to. After this point, we know that x is unique. We want to
+  // preserve that property.
+  x.append(5)
+  
+  // We create a new variable y so we can write an algorithm where we may
+  // change the value of y (causing a COW copy of the buffer shared with x).
+  var y = x
+  longAlgorithmUsing(&y)
+  // We no longer use y after this point, so move it when we pass it off to
+  // the last use.
+  consumeFinalY(move(y))
+
+  // x will be unique again here.
+  x.append(7)
+}
+```
+
+This addresses both of the motivating issues above: `move` guarantees the
+lifetime of `y` ends at the given point, allowing the compiler to generate
+code to clean up or transfer ownership of `y` without relying on optimization.
+Furthermore, if a future maintainer modifies the code in a way that extends
+the lifetime of `y` past the expected point, then the compiler will raise an
+error. For instance, if we try:
+
+```
+func test() {
+  var x: [Int] = getArray()
+  
+  // x is appended to. After this point, we know that x is unique. We want to
+  // preserve that property.
+  x.append(5)
+  
+  // We create a new variable y so we can write an algorithm where we may
+  // change the value of y (causing a COW copy of the buffer shared with x).
+  var y = x
+  longAlgorithmUsing(&y)
+  // We think we no longer use y after this point...
+  consumeFinalY(move(y))
+
+  // ...and x will be unique again here...
+  x.append(7)
+
+  // ...but this additional use of y snuck in:
+  useYAgain(y)
+}
+```
+
+In this case, we get the following output from the compiler as expected:
+
+```
+test.swift:10:7: error: 'y' used after being moved
+  var y = x
+      ^
+test.swift:13:17: note: move here
+  consumeFinalY(move(y))
+                ^
+test.swift:19:13: note: use here
+  useYAgain(y)
+            ^
+```
+
+Note how the compiler gives all of the information that we need to resolve
```

Comment:
Strike this paragraph.  It only talks about the quality of your prototype implementation, which is not the issue.  This document should focus on the language feature and the need for it.

(Personally, I would cut out most of the previous example as well; only the first error message is truly relevant.)

Reply:
On further reflection, this paragraph really needs to go.